File: apps/web/lib/crypto/kek-holder.ts
'use client';

/**
 * In-memory-only holder for the current tab's local-storage KEK — derived once at
 * unlock (login/invite-redeem/device-link), used to wrap/unwrap sessions for the
 * rest of that page's lifetime, never itself persisted (docs/05-crypto-architecture.md#local-key-storage:
 * "the KEK itself lives only in memory for the session"). A page reload or tab close
 * clears this by construction — the user unlocks again via `unlockLocalIdentity`.
 */
let currentKek: Uint8Array | null = null;
let currentIdentity: import('@comm/crypto').IdentityKeyPair | null = null;

export function setUnlockedIdentity(kek: Uint8Array, identity: import('@comm/crypto').IdentityKeyPair): void {
  currentKek = kek;
  currentIdentity = identity;
}

export function getCurrentKek(): Uint8Array | null {
  return currentKek;
}

export function getCurrentIdentity(): import('@comm/crypto').IdentityKeyPair | null {
  return currentIdentity;
}

export function clearUnlockedIdentity(): void {
  if (currentKek) currentKek.fill(0);
  currentKek = null;
  currentIdentity = null;
}

File: apps/mobile/lib/crypto/kek_holder.dart
/// In-memory-only holder for the current process's local-storage KEK — direct port
/// of `apps/web/lib/crypto/kek-holder.ts`. Derived once at unlock (login/invite
/// redeem/device link), used to wrap/unwrap sessions for the rest of the app's
/// lifetime, never itself persisted: "the KEK lives only in memory for the session."
/// An app kill/restart clears this by construction — the user unlocks again via
/// `unlockLocalIdentity` (crypto/local_identity.dart), same as a page reload on web.
library;

import 'dart:typed_data';
import 'identity/keys.dart';
import 'primitives.dart';

Uint8List? _currentKek;
IdentityKeyPair? _currentIdentity;

void setUnlockedIdentity(Uint8List kek, IdentityKeyPair identity) {
  _currentKek = kek;
  _currentIdentity = identity;
}

Uint8List? getCurrentKek() => _currentKek;

IdentityKeyPair? getCurrentIdentity() => _currentIdentity;

void clearUnlockedIdentity() {
  final kek = _currentKek;
  if (kek != null) wipe(kek);
  _currentKek = null;
  _currentIdentity = null;
}
